### PR TITLE
[NN] Enhance EGATConv branch

### PR DESCRIPTION
**Pull request by rijulizer**
## Description
Extending dgl.nn.EGATConv module to handle bipartite graphs, an effort towards making it compatible with heterogeneous graphs.
## Link to the discussion: 
https://discuss.dgl.ai/t/handling-edge-features-in-heterographconv-wrapper-for-heterogeneous-graphs/2935/14?u=rijulizer

### CODE REVIEW

File: python/dgl/nn/pytorch/conv/egatconv.py [hunk 1]
```diff
@@ -27,8 +28,14 @@ class EGATConv(nn.Module):
 
     Parameters
     ----------
-    in_node_feats : int
-        Input node feature size :math:`h_{i}`.
+    in_node_feats : int, or pair of ints
+        Input feature size; i.e, the number of dimensions of :math:`h_{i}`.
+        GATConv can be applied on homogeneous graph and unidirectional
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egatconv.py:33]: `GATConv` -> `EGATConv`?

File: python/dgl/nn/pytorch/conv/egatconv.py [hunk 2]
```diff
@@ -61,7 +69,36 @@ class EGATConv(nn.Module):
...
+    >>> efeats = th.tensor(np.random.rand(5, 15).astype(np.float32))
+
+    >>> in_node_feats = (25,30)
+    >>> in_edge_feats = 15
+    >>> out_node_feats = 10
+    >>> out_edge_feats = 5
+    >>> num_heads = 3
+
+    >>> egat_model =  EGATConv(in_node_feats,
+    ...                         in_edge_feats,
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egatconv.py:90]: Could you fix the indent here?

File: python/dgl/nn/pytorch/conv/egatconv.py [hunk 3]
```diff
@@ -106,11 +157,14 @@ def forward(self, graph, nfeats, efeats, get_attention=False):
         ----------
         graph : DGLGraph
             The graph.
-        nfeats : torch.Tensor
-            The input node feature of shape :math:`(N, D_{in})`
+        nfeat : torch.Tensor or pair of torch.Tensor
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egatconv.py:160]: Does this module support features of dimensionality 3 or more?

File: python/dgl/nn/pytorch/conv/egatconv.py [hunk 4]
```diff
@@ -106,11 +157,14 @@ def forward(self, graph, nfeats, efeats, get_attention=False):
...
+            If a torch.Tensor is given, the input feature of shape :math:`(N, *, D_{in})` 
             where:
-                :math:`D_{in}` is size of input node feature,
+                :math:`D_{in}` is size of input node feature, 
                 :math:`N` is the number of nodes.
+            If a pair of torch.Tensor is given, the pair must contain two tensors of shape
+                :math:`(N_{in}, *, D_{in_{src}})` and
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egatconv.py:166]: Does this module support feature tensors of dimensionality 3 or more?

**Comment by mufeili** [comment]
Could you add a unit test for the case of unidirectional bipartite graphs?

File: python/dgl/nn/pytorch/conv/egatconv.py (hunk 1 above)
Comment (rijulizer) [reply python/dgl/nn/pytorch/conv/egatconv.py:33]: Yes, I will fix it

File: python/dgl/nn/pytorch/conv/egatconv.py (hunk 2 above)
Comment (rijulizer) [reply python/dgl/nn/pytorch/conv/egatconv.py:90]: ok.

File: python/dgl/nn/pytorch/conv/egatconv.py (hunk 4 above)
Comment (rijulizer) [reply python/dgl/nn/pytorch/conv/egatconv.py:166]: Interesting, neither the existing version nor my update supports it. Can be an interesting update.

**Comment by rijulizer** [comment]
> Could you add a unit test for the case of unidirectional bipartite graphs?

Where do you want me to add it? in the docs?

**Comment by rijulizer** [comment]
> Could you add a unit test for the case of unidirectional bipartite graphs?

Added unit test and modified the existing one for homogeneous graph.

**Review by mufeili — APPROVED** [review]
LGTM. Thanks.

**Comment by mufeili** [comment]
Lint check failed

```
python/dgl/nn/pytorch/conv/egatconv.py:55:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/egatconv.py:71:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/egatconv.py:92:0: C0301: Line too long (102/100) (line-too-long)

python/dgl/nn/pytorch/conv/egatconv.py:96:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/egatconv.py:111:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/egatconv.py:156:86: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/egatconv.py:158:61: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/egatconv.py:161:50: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/egatconv.py:221:91: C0303: Trailing whitespace (trailing-whitespace)
```

**Pull request merged**